webapp/src/Controller/Jury/ExecutableController.php

```diff
@@ -206,14 +156,81 @@ public function indexAction(Request $request)
         ]);
     }
 
+    /**
+     * @Route("/add", name="jury_executable_add")
+     * @IsGranted("ROLE_ADMIN")
+     * @param Request $request
+     * @return Response
+     */
+    public function addAction(Request $request) : Response
+    {
+        $data = [];
+        $form = $this->createForm(ExecutableUploadType::class, $data);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()){
+            $propertyFile = 'domjudge-executable.ini';
+            $data         = $form->getData();
+            /** @var UploadedFile[] $archives */
+            $archives = $data['archives'];
+            $id       = null;
+            foreach ($archives as $archive) {
+                $zip         = $this->dj->openZipFile($archive->getRealPath());
+                $filename    = $archive->getClientOriginalName();
+                $id          = substr($filename, 0, strlen($filename) - strlen(".zip"));
+                if ( ! preg_match ('#^[a-z0-9_-]+$#i', $id) ) {
+                    throw new InvalidArgumentException(sprintf("File base name '%s' must contain only alphanumerics", $id));
+                }
+                $description = $id;
+                $type        = $data['type'];
+
+                $propertyData = $zip->getFromName($propertyFile);
+                if ($propertyData !== false) {
+                    $ini_array = parse_ini_string($propertyData);
+                } else {
+                    $ini_array = [];
+                }
+                if (!empty($ini_array)) {
+                    $id          = $ini_array['execid'];
+                    $description = $ini_array['description'];
+                    $type        = $ini_array['type'];
+                }
+
+                $immutableExecutable = $this->dj->createImmutableExecutable($zip, $propertyFile);
+                $executable = new Executable();
+                $executable
+                    ->setExecid($id)
+                    ->setDescription($description)
+                    ->setType($type)
+                    ->setImmutableExecutable($immutableExecutable);
+                $this->em->persist($executable);
+
+                $zip->close();
+
+                $this->dj->auditlog('executable', $id, 'upload zip', $archive->getClientOriginalName());
+            }
+
+            $this->em->flush();
+
+            if (count($archives) === 1) {
+                return $this->redirectToRoute('jury_executable', ['execId' => $id]);
+            } else {
+                return $this->redirectToRoute('jury_executables');
+            }
+        }
+
+        return $this->render('jury/executable_add.html.twig', [
+            'form' => $form->createView(),
+        ]);
+    }
+
     /**
      * @Route("/{execId}", name="jury_executable")
      * @param Request $request
```

Comment:
Here as well. More below which you can find

Reply:
Should now be removed together with PHPDoc parts which are just trivial